src/Unused/CLI/SearchWithProgress.hs

```diff
@@ -1,40 +1,45 @@
+{-# LANGUAGE FlexibleContexts #-}
+
 module Unused.CLI.SearchWithProgress
     ( searchWithProgressBar
     ) where
 
 import Control.Monad.State
+import Data.Bifunctor (first)
 import System.ProgressBar
 import Unused.TermSearch (search)
 
+newtype Progress a = Progress a
+
+instance Functor Progress where
+    fmap f (Progress s) = Progress $ f s
+
+type ProgressState = Progress (Integer, Integer)
+
 searchWithProgressBar :: [String] -> IO [String]
 searchWithProgressBar terms =
-    (concat . fst) <$> runStateT (performSearch $ length terms) terms
-
-performSearch :: Int -> StateT [String] IO [[String]]
-performSearch total = do
-    currentTerm <- getSearchTerm
-    searchResults <- liftIO $ search currentTerm
+    concat . fst <$> runStateT (mapM performSearch terms) progress
+  where
+    progress = Progress (0, toInteger $ length terms)
 
-    remainingTerms <- get
-    let remaining = length remainingTerms
+performSearch :: String -> StateT ProgressState IO [String]
+performSearch term = do
+    results <- liftIO $ search term
 
-    liftIO $ printProgressBar (total - remaining) total
+    s <- incrementCount
+    liftIO $ printProgressBar s
 
-    if remaining > 0
-        then do
-            res <- performSearch total
-            return $ searchResults:res
-        else return [searchResults]
+    return results
 
-getSearchTerm :: StateT [String] IO String
-getSearchTerm = do
-    (x:xs) <- get
-    put xs
-    return x
+incrementCount :: (MonadState ProgressState m) => m ProgressState
+incrementCount = do
+    s <- get
+    put $ first (+ 1) <$> s
+    return s
 
-printProgressBar :: Int -> Int -> IO ()
-printProgressBar complete total = do
+printProgressBar :: ProgressState -> IO ()
```

Comment:
you can do better here by making this a monad action `printProgressBar :: StateT ProgressState IO ()`